### PR TITLE
Fix BPF map data race conditions

### LINKED ISSUE #10990: DATA RACE: m.fd unsafely accessed
It seems this was introduced by 47e01ebd465d3f9f2d4534133be10b0f13ce8053

source 1000f7023a46f2ff0c29b3b155c294942b5c4fd7

```
2020-04-15T09:18:45.371539261Z WARNING: DATA RACE
2020-04-15T09:18:45.371643044Z Write at 0x00c0008ec7d0 by goroutine 21:
2020-04-15T09:18:45.371650058Z   github.com/cilium/cilium/pkg/bpf.(*Map).Open()
2020-04-15T09:18:45.371652544Z       /go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:514 +0x1b6
2020-04-15T09:18:45.371654925Z   github.com/cilium/cilium/pkg/maps/nat.(*Map).Open()
2020-04-15T09:18:45.371657305Z       <autogenerated>:1 +0x43
2020-04-15T09:18:45.371679261Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC6()
2020-04-15T09:18:45.371686445Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:289 +0x47a
2020-04-15T09:18:45.371748303Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC()
2020-04-15T09:18:45.371755107Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:463 +0x93
2020-04-15T09:18:45.37175759Z   github.com/cilium/cilium/pkg/maps/ctmap.GC()
2020-04-15T09:18:45.371759998Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:480 +0x5b
2020-04-15T09:18:45.371785982Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).garbageCollectConntrack()
2020-04-15T09:18:45.37179223Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:945 +0x867
2020-04-15T09:18:45.371843948Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).scrubIPsInConntrackTableLocked()
2020-04-15T09:18:45.371863565Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:950 +0x19e
2020-04-15T09:18:45.371869155Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).leaveLocked()
2020-04-15T09:18:45.371871564Z       /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1014 +0x3c7
2020-04-15T09:18:45.371904738Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Delete()
2020-04-15T09:18:45.371910913Z       /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:2078 +0x3d1
2020-04-15T09:18:45.371967251Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).deleteEndpointQuiet()
2020-04-15T09:18:45.371974058Z       /go/src/github.com/cilium/cilium/daemon/cmd/endpoint.go:481 +0xc3
2020-04-15T09:18:45.37197647Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints.func2()
2020-04-15T09:18:45.371978833Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:289 +0x36
2020-04-15T09:18:45.37198117Z 
2020-04-15T09:18:45.372043756Z Previous read at 0x00c0008ec7d0 by goroutine 177:
2020-04-15T09:18:45.372094279Z   github.com/cilium/cilium/pkg/bpf.(*Map).Close()
2020-04-15T09:18:45.372100286Z       /go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:527 +0xe5
2020-04-15T09:18:45.372162319Z   github.com/cilium/cilium/pkg/maps/nat.(*Map).Close()
2020-04-15T09:18:45.372169095Z       <autogenerated>:1 +0x43
2020-04-15T09:18:45.37217736Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC6()
2020-04-15T09:18:45.372179929Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:350 +0x44e
2020-04-15T09:18:45.372190316Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC()
2020-04-15T09:18:45.372196534Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:463 +0x93
2020-04-15T09:18:45.372258732Z   github.com/cilium/cilium/pkg/maps/ctmap.GC()
2020-04-15T09:18:45.372265589Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:480 +0x5b
2020-04-15T09:18:45.372268079Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).garbageCollectConntrack()
2020-04-15T09:18:45.372283651Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:945 +0x867
2020-04-15T09:18:45.37228857Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).scrubIPsInConntrackTableLocked()
2020-04-15T09:18:45.372291117Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:950 +0x19e
2020-04-15T09:18:45.372319191Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).leaveLocked()
2020-04-15T09:18:45.37232542Z       /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1014 +0x3c7
2020-04-15T09:18:45.372380205Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Delete()
2020-04-15T09:18:45.372386971Z       /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:2078 +0x3d1
2020-04-15T09:18:45.372389434Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).deleteEndpointQuiet()
2020-04-15T09:18:45.372391715Z       /go/src/github.com/cilium/cilium/daemon/cmd/endpoint.go:481 +0xc3
2020-04-15T09:18:45.372394007Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints.func2()
2020-04-15T09:18:45.372396335Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:289 +0x36
2020-04-15T09:18:45.372418403Z 
2020-04-15T09:18:45.372475712Z Goroutine 21 (running) created at:
2020-04-15T09:18:45.372526749Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints()
2020-04-15T09:18:45.372533535Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:283 +0x985
2020-04-15T09:18:45.372535979Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).initRestore()
2020-04-15T09:18:45.372538341Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:355 +0xf3
2020-04-15T09:18:45.372558552Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-04-15T09:18:45.372564554Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1237 +0xa72
2020-04-15T09:18:45.37264223Z   github.com/cilium/cilium/daemon/cmd.NewDaemon()
2020-04-15T09:18:45.372649541Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:508 +0x2fa5
2020-04-15T09:18:45.37269938Z   github.com/cilium/cilium/daemon/cmd.NewDaemon()
2020-04-15T09:18:45.372706515Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:411 +0x463e
2020-04-15T09:18:45.372714033Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-04-15T09:18:45.372716517Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1194 +0x345
2020-04-15T09:18:45.372719807Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2020-04-15T09:18:45.372722236Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:107 +0x108
2020-04-15T09:18:45.372749505Z   github.com/cilium/cilium/daemon/cmd.initEnv()
2020-04-15T09:18:45.37275583Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1073 +0x2c56
2020-04-15T09:18:45.37280784Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2020-04-15T09:18:45.372814649Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:105 +0xee
2020-04-15T09:18:45.372817109Z   github.com/spf13/cobra.(*Command).execute()
2020-04-15T09:18:45.37281954Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:830 +0x8e0
2020-04-15T09:18:45.372830709Z   github.com/spf13/cobra.(*Command).ExecuteC()
2020-04-15T09:18:45.37283671Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:914 +0x41a
2020-04-15T09:18:45.372913775Z   github.com/spf13/cobra.(*Command).Execute()
2020-04-15T09:18:45.372920436Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:864 +0x251
2020-04-15T09:18:45.372922943Z   github.com/cilium/cilium/daemon/cmd.Execute()
2020-04-15T09:18:45.372925219Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:138 +0x232
2020-04-15T09:18:45.372927556Z   main.main()
2020-04-15T09:18:45.372929845Z       /go/src/github.com/cilium/cilium/daemon/main.go:22 +0x2f
2020-04-15T09:18:45.372946212Z 
2020-04-15T09:18:45.372983536Z Goroutine 177 (running) created at:
2020-04-15T09:18:45.373034377Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints()
2020-04-15T09:18:45.373040112Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:283 +0x985
2020-04-15T09:18:45.373103446Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).initRestore()
2020-04-15T09:18:45.373109511Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:355 +0xf3
2020-04-15T09:18:45.373156605Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-04-15T09:18:45.373163529Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1237 +0xa72
2020-04-15T09:18:45.373165968Z   github.com/cilium/cilium/daemon/cmd.NewDaemon()
2020-04-15T09:18:45.373168253Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:508 +0x2fa5
2020-04-15T09:18:45.373188586Z   github.com/cilium/cilium/daemon/cmd.NewDaemon()
2020-04-15T09:18:45.373194825Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:411 +0x463e
2020-04-15T09:18:45.373240409Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-04-15T09:18:45.373246257Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1194 +0x345
2020-04-15T09:18:45.373298723Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2020-04-15T09:18:45.373305431Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:107 +0x108
2020-04-15T09:18:45.373307858Z   github.com/cilium/cilium/daemon/cmd.initEnv()
2020-04-15T09:18:45.373310156Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1073 +0x2c56
2020-04-15T09:18:45.373326056Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2020-04-15T09:18:45.373332342Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:105 +0xee
2020-04-15T09:18:45.373379611Z   github.com/spf13/cobra.(*Command).execute()
2020-04-15T09:18:45.373385493Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:830 +0x8e0
2020-04-15T09:18:45.373430178Z   github.com/spf13/cobra.(*Command).ExecuteC()
2020-04-15T09:18:45.373436895Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:914 +0x41a
2020-04-15T09:18:45.37343934Z   github.com/spf13/cobra.(*Command).Execute()
2020-04-15T09:18:45.373441636Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:864 +0x251
2020-04-15T09:18:45.373468811Z   github.com/cilium/cilium/daemon/cmd.Execute()
2020-04-15T09:18:45.373474804Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:138 +0x232
2020-04-15T09:18:45.373522068Z   main.main()
2020-04-15T09:18:45.373528761Z       /go/src/github.com/cilium/cilium/daemon/main.go:22 +0x2f
```

```
2020-04-15T09:03:37.916390453Z ==================
2020-04-15T09:03:37.916555933Z WARNING: DATA RACE
2020-04-15T09:03:37.916945291Z Read at 0x00c000519450 by goroutine 246:
2020-04-15T09:03:37.917034918Z   github.com/cilium/cilium/pkg/bpf.(*Map).Open()
2020-04-15T09:03:37.917042636Z       /go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:499 +0xc3
2020-04-15T09:03:37.917094615Z   github.com/cilium/cilium/pkg/maps/nat.(*Map).Open()
2020-04-15T09:03:37.917100895Z       <autogenerated>:1 +0x43
2020-04-15T09:03:37.91714663Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC6()
2020-04-15T09:03:37.917152583Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:289 +0x47a
2020-04-15T09:03:37.917214346Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC()
2020-04-15T09:03:37.91722028Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:463 +0x93
2020-04-15T09:03:37.917265303Z   github.com/cilium/cilium/pkg/maps/ctmap.GC()
2020-04-15T09:03:37.917271196Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:480 +0x5b
2020-04-15T09:03:37.917332428Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).garbageCollectConntrack()
2020-04-15T09:03:37.917338575Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:945 +0x867
2020-04-15T09:03:37.917383313Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).scrubIPsInConntrackTableLocked()
2020-04-15T09:03:37.917389293Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:950 +0x19e
2020-04-15T09:03:37.917450252Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).leaveLocked()
2020-04-15T09:03:37.917456291Z       /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1014 +0x3c7
2020-04-15T09:03:37.917520822Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Delete()
2020-04-15T09:03:37.917527118Z       /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:2078 +0x3d1
2020-04-15T09:03:37.91756484Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).deleteEndpointQuiet()
2020-04-15T09:03:37.917570071Z       /go/src/github.com/cilium/cilium/daemon/cmd/endpoint.go:481 +0xc3
2020-04-15T09:03:37.917633612Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints.func2()
2020-04-15T09:03:37.917639823Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:289 +0x36
2020-04-15T09:03:37.917686571Z 
2020-04-15T09:03:37.91778602Z Previous write at 0x00c000519450 by goroutine 273:
2020-04-15T09:03:37.917849044Z   github.com/cilium/cilium/pkg/bpf.(*Map).Close()
2020-04-15T09:03:37.917855257Z       /go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:529 +0x181
2020-04-15T09:03:37.917901233Z   github.com/cilium/cilium/pkg/maps/nat.(*Map).Close()
2020-04-15T09:03:37.91790644Z       <autogenerated>:1 +0x43
2020-04-15T09:03:37.917963862Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC6()
2020-04-15T09:03:37.917969869Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:350 +0x44e
2020-04-15T09:03:37.918027212Z   github.com/cilium/cilium/pkg/maps/ctmap.doGC()
2020-04-15T09:03:37.918033172Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:463 +0x93
2020-04-15T09:03:37.918131984Z   github.com/cilium/cilium/pkg/maps/ctmap.GC()
2020-04-15T09:03:37.918138754Z       /go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:480 +0x5b
2020-04-15T09:03:37.918176289Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).garbageCollectConntrack()
2020-04-15T09:03:37.918181477Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:945 +0x867
2020-04-15T09:03:37.918248349Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).scrubIPsInConntrackTableLocked()
2020-04-15T09:03:37.918254889Z       /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:950 +0x19e
2020-04-15T09:03:37.918292136Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).leaveLocked()
2020-04-15T09:03:37.91829726Z       /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1014 +0x3c7
2020-04-15T09:03:37.918353636Z   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Delete()
2020-04-15T09:03:37.91835977Z       /go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:2078 +0x3d1
2020-04-15T09:03:37.918410689Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).deleteEndpointQuiet()
2020-04-15T09:03:37.918416372Z       /go/src/github.com/cilium/cilium/daemon/cmd/endpoint.go:481 +0xc3
2020-04-15T09:03:37.918459902Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints.func2()
2020-04-15T09:03:37.918465628Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:289 +0x36
2020-04-15T09:03:37.918524094Z 
2020-04-15T09:03:37.918571222Z Goroutine 246 (running) created at:
2020-04-15T09:03:37.918633978Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints()
2020-04-15T09:03:37.918640065Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:283 +0x985
2020-04-15T09:03:37.918703832Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).initRestore()
2020-04-15T09:03:37.91870999Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:355 +0xf3
2020-04-15T09:03:37.918746825Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-04-15T09:03:37.91875192Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1237 +0xa72
2020-04-15T09:03:37.918815873Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-04-15T09:03:37.918822064Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1194 +0x345
2020-04-15T09:03:37.918858913Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2020-04-15T09:03:37.918864073Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:107 +0x108
2020-04-15T09:03:37.918924113Z   github.com/spf13/cobra.(*Command).execute()
2020-04-15T09:03:37.918930054Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:830 +0x8e0
2020-04-15T09:03:37.918966578Z   github.com/spf13/cobra.(*Command).ExecuteC()
2020-04-15T09:03:37.918971688Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:914 +0x41a
2020-04-15T09:03:37.919035095Z   github.com/spf13/cobra.(*Command).Execute()
2020-04-15T09:03:37.919041584Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:864 +0x251
2020-04-15T09:03:37.919079067Z   github.com/cilium/cilium/daemon/cmd.Execute()
2020-04-15T09:03:37.919084308Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:138 +0x232
2020-04-15T09:03:37.919175156Z   main.main()
2020-04-15T09:03:37.919181411Z       /go/src/github.com/cilium/cilium/daemon/main.go:22 +0x2f
2020-04-15T09:03:37.919228051Z 
2020-04-15T09:03:37.919295665Z Goroutine 273 (running) created at:
2020-04-15T09:03:37.919336421Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).regenerateRestoredEndpoints()
2020-04-15T09:03:37.919354346Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:283 +0x985
2020-04-15T09:03:37.919399621Z   github.com/cilium/cilium/daemon/cmd.(*Daemon).initRestore()
2020-04-15T09:03:37.919405379Z       /go/src/github.com/cilium/cilium/daemon/cmd/state.go:355 +0xf3
2020-04-15T09:03:37.919463474Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-04-15T09:03:37.919469385Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1237 +0xa72
2020-04-15T09:03:37.919513432Z   github.com/cilium/cilium/daemon/cmd.runDaemon()
2020-04-15T09:03:37.91951925Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1194 +0x345
2020-04-15T09:03:37.919579754Z   github.com/cilium/cilium/daemon/cmd.glob..func1()
2020-04-15T09:03:37.919585506Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:107 +0x108
2020-04-15T09:03:37.91963021Z   github.com/spf13/cobra.(*Command).execute()
2020-04-15T09:03:37.919636073Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:830 +0x8e0
2020-04-15T09:03:37.919697031Z   github.com/spf13/cobra.(*Command).ExecuteC()
2020-04-15T09:03:37.919702862Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:914 +0x41a
2020-04-15T09:03:37.919747125Z   github.com/spf13/cobra.(*Command).Execute()
2020-04-15T09:03:37.919752894Z       /go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:864 +0x251
2020-04-15T09:03:37.919813445Z   github.com/cilium/cilium/daemon/cmd.Execute()
2020-04-15T09:03:37.919819287Z       /go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:138 +0x232
2020-04-15T09:03:37.919864526Z   main.main()
2020-04-15T09:03:37.919870348Z       /go/src/github.com/cilium/cilium/daemon/main.go:22 +0x2f
2020-04-15T09:03:37.919938385Z ==================
```

### PR DESCRIPTION
**Pull request by tgraf**
Fixes: #10990

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
restart-gke

**Comment by tgraf** [comment]
test-gke

File: pkg/bpf/map_linux.go [hunk 1]
```diff
@@ -226,17 +229,25 @@ func (m *Map) WithCache() *Map {
...
 }
 
-func (m *Map) GetFd() int {
-	return m.fd
+// GetFd returns the file descriptor of the BPF map
+func (m *Map) GetFd() (fd int) {
+	m.lock.RLock()
+	fd = m.fd
+	m.lock.RUnlock()
+	return
```
Comment (joestringer) [inline pkg/bpf/map_linux.go]: ICYMI, the approach below is now optimized by the Go compiler to "do the right thing" (efficiently):

```
m.RLock()
defer m.RUnlock()
return m.fd
```

(I don't care that much if it's used here or not so feel free to disregard this comment)

File: pkg/bpf/map_linux.go [hunk 2]
```diff
@@ -822,14 +880,16 @@ func (m *Map) Update(key MapKey, value MapValue) error {
 		}
 	}()
 
-	if err = m.Open(); err != nil {
+	if err = m.openAndLock(); err != nil {
```
Comment (joestringer) [inline pkg/bpf/map_linux.go]: In this error condition, we will not have the lock and we will still perform the defer() immediately above, including the potential modification of `m.cache`.

File: pkg/bpf/map_linux.go [hunk 3]
```diff
@@ -864,15 +924,13 @@ func (m *Map) deleteCacheEntry(key MapKey, err error) {
...
 // Delete deletes the map entry corresponding to the given key.
 func (m *Map) Delete(key MapKey) error {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-
 	var err error
 	defer m.deleteCacheEntry(key, err)
 
-	if err = m.Open(); err != nil {
+	if err = m.openAndLock(); err != nil {
```
Comment (joestringer) [inline pkg/bpf/map_linux.go]: Similar usage of `m.cache` as the above comment, via `m.deleteCacheEntry()`.

File: pkg/bpf/map_linux.go [hunk 4]
```diff
@@ -1054,7 +1054,9 @@ func (m *Map) GetModel() *models.BPFMap {
 func (m *Map) resolveErrors(ctx context.Context) error {
 	started := time.Now()
 
-	m.lock.Lock()
+	if err := m.openAndLock(); err != nil {
```
Comment (joestringer) [inline pkg/bpf/map_linux.go:1060]: TIL open / close don't need to be balanced; maps can be opened any number of times but should be closed once.

LGTM.

**Review by joestringer — CHANGES_REQUESTED** [review]
Looks like a couple of `m.cache` usages without holding the lock were missed, PTAL.

**Comment by tgraf** [comment]
test-me-please

**Comment by aanm** [comment]
The GKE test failures seem to be related as they are touching bpf maps? `bpf-map-sync-cilium_ipcache           never          4s ago       9       4 map sync errors`

File: pkg/bpf/map_linux.go [hunk 5]
```diff
@@ -92,10 +91,16 @@ type cacheEntry struct {
...
+	// be held
+	name string
+
 	MapInfo
 	fd   int
-	name string
 	path string
+
+	// lock protects all fields of this structure and serializes calls to
+	// all Open() variants
```
Comment (aanm) [inline pkg/bpf/map_linux.go:103]: Conventionally, a lock is on top of all fields that it protects so this should be on line 97 to be more explicit.

File: pkg/bpf/map_linux.go [hunk 6]
```diff
@@ -549,18 +583,20 @@ type MapValidator func(path string) (bool, error)
...
 // corruption.
+//
+// It is safe to perform BPF map operations from within the callback
 func (m *Map) DumpWithCallback(cb DumpCallback) error {
-	m.lock.RLock()
+	if err := m.openAndRLock(); err != nil {
+		return err
+	}
+	// Even if the lock is temporarily released, the lock is *always* held
+	// when the function exits
```
Comment (aanm) [inline pkg/bpf/map_linux.go:593]: Looking at the previous code it seems to me that some call backs were only called after `DumpWithCallback` was returned? Is my assessment correct?

**Review by aanm — CHANGES_REQUESTED** [review]
Also there's potential deadlock for the `m.Flush()` call.

```
POTENTIAL DEADLOCK: Recursive locking:
current goroutine 24 lock 0xc00026e078
../../lock/lock_debug.go:62 lock.(*internalRWMutex).Lock { i.RWMutex.Lock() } <<<<<
../../bpf/map_linux.go:930 bpf.(*Map).Delete { m.lock.Lock() }
ctmap.go:377 ctmap.purgeCtEntry4 { err := m.Delete(key) }
ctmap.go:414 ctmap.doGC4.func1 { err := purgeCtEntry4(m, currentKey4Global, natMap) }
../../bpf/map_linux.go:791 bpf.(*Map).DumpReliablyWithCallback { cb(mk, mv) }
ctmap.go:451 ctmap.doGC4 { stats.dumpError = m.DumpReliablyWithCallback(filterCallback, stats.DumpStats) }
ctmap.go:488 ctmap.doGC { return int(doGC4(m, filter).deleted) }
ctmap.go:509 ctmap.(*Map).Flush { return doGC(m, &GCFilter{ }
ctmap.go:509 ctmap.(*Map).Flush { return doGC(m, &GCFilter{ }
/usr/local/go/src/reflect/value.go:460 reflect.Value.call { call(frametype, fn, args, uint32(frametype.size), uint32(retOffset)) }
/usr/local/go/src/reflect/value.go:321 reflect.Value.Call { return v.call("Call", in) }
../../../vendor/gopkg.in/check.v1/check.go:792 check%2ev1.(*suiteRunner).forkTest.func1 { c.method.Call([]reflect.Value{reflect.ValueOf(c)}) }
../../../vendor/gopkg.in/check.v1/check.go:675 check%2ev1.(*suiteRunner).forkCall.func1 { dispatcher(c) }

Previous place where the lock was grabbed (same goroutine)
../../lock/lock_debug.go:78 bpf.(*Map).openAndRLock { i.RWMutex.RLock() } <<<<<
../../bpf/map_linux.go:503 bpf.(*Map).openAndRLock { m.lock.RLock() }
../../bpf/map_linux.go:694 bpf.(*Map).DumpReliablyWithCallback { if err := m.openAndRLock(); err != nil { }
ctmap.go:451 ctmap.doGC4 { stats.dumpError = m.DumpReliablyWithCallback(filterCallback, stats.DumpStats) }
ctmap.go:488 ctmap.doGC { return int(doGC4(m, filter).deleted) }
ctmap.go:509 ctmap.(*Map).Flush { return doGC(m, &GCFilter{ }
ctmap.go:509 ctmap.(*Map).Flush { return doGC(m, &GCFilter{ }
/usr/local/go/src/reflect/value.go:460 reflect.Value.call { call(frametype, fn, args, uint32(frametype.size), uint32(retOffset)) }
/usr/local/go/src/reflect/value.go:321 reflect.Value.Call { return v.call("Call", in) }
../../../vendor/gopkg.in/check.v1/check.go:792 check%2ev1.(*suiteRunner).forkTest.func1 { c.method.Call([]reflect.Value{reflect.ValueOf(c)}) }
../../../vendor/gopkg.in/check.v1/check.go:675 check%2ev1.(*suiteRunner).forkCall.func1 { dispatcher(c) }

Other goroutines holding locks:
goroutine 24 lock 0x12a4340
../../lock/lock_debug.go:91 lock.(*internalMutex).Lock { i.Mutex.Lock() } <<<<<
ctmap.go:450 ctmap.doGC4 { globalDeleteLock[m.mapType].Lock() }
ctmap.go:488 ctmap.doGC { return int(doGC4(m, filter).deleted) }
ctmap.go:509 ctmap.(*Map).Flush { return doGC(m, &GCFilter{ }
ctmap.go:509 ctmap.(*Map).Flush { return doGC(m, &GCFilter{ }
/usr/local/go/src/reflect/value.go:460 reflect.Value.call { call(frametype, fn, args, uint32(frametype.size), uint32(retOffset)) }
/usr/local/go/src/reflect/value.go:321 reflect.Value.Call { return v.call("Call", in) }
../../../vendor/gopkg.in/check.v1/check.go:792 check%2ev1.(*suiteRunner).forkTest.func1 { c.method.Call([]reflect.Value{reflect.ValueOf(c)}) }
../../../vendor/gopkg.in/check.v1/check.go:675 check%2ev1.(*suiteRunner).forkCall.func1 { dispatcher(c) }
```

**Comment by joestringer** [comment]
We discussed offline scoping this out of the next v1.7 release, so moving to prio/high.

**Comment by tgraf** [comment]
Closed in favor of #11812

**Pull request closed without merging**